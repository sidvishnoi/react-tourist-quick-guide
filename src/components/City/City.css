.City {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  background: linear-gradient(to right, #ff8a00, #da1b60);
  margin: 1em auto;
  padding: 0.2em;
  transition: all 0.2s ease;
}

.City:hover {
  border-image: linear-gradient(to right, #ff8a00, #da1b60) 1 1;
}

.City > div {
  flex-grow: 1;
}

.City .widget {
  background: #fff;
}

.City .meta {
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  background: transparent;
  /* background-image: linear-gradient(to right, #ff8a00, #da1b60); */
  width: 200px;
}

.City .meta .buttons {
  display: flex;
  flex-direction: column;
  width: 30px;
}

.City .meta button {
  background: rgba(255, 255, 255, 0.2);
  border: none;
  padding: 0.5rem;
  margin: 0.1rem;
  color: crimson;
}

.City .meta button:hover,
.City .meta button:focus {
  background: crimson;
  color: #fff;
  cursor: pointer;
}

.City .meta h3 {
  font-size: 2rem;
  text-align: center;
  padding: 0.2rem 0.5rem;
  color: #000;
}

.City .meta .distance {
  text-align: right;
  padding: 1rem;
  font-size: 1.2rem;
}

@media print {
  .City {
    background: #fff;
    border-bottom: 2px solid #000;
  }

  .City .meta {
    justify-content: flex-start;
    width: auto;
    max-width: 200px;
  }

  .City .meta .buttons {
    display: none;
  }

  .City .meta h3,
  .City .meta .distance {
    padding: 0;
    text-align: left;
  }
}
