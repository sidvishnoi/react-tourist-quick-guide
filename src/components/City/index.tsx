import * as React from 'react';
import styled from 'styled-components';

import CityMeta from '../CityMeta';
import Places from '../Places';
import Weather, { WeatherProps } from '../Weather';
import Widget from '../Widget';

export interface CityProps {
  name: string;
  weather: { state: string; data: WeatherProps };
  places: { state: string; data: { name: string; link: string }[] };
  distance: { state: string; data: number };
}

interface CityPropsFromDispatch {
  destroyer?: (name: string) => void;
  mover?: (name: string) => void;
}

const City = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-start;
  margin: 1em auto;
  border: 2px solid #000;
  max-width: 900px;

  > div {
    max-width: 300px;
    flex-grow: 1;
  }

  @media screen and (max-width: 700px) {
    > div {
      max-width: 100%;
    }
  }

  @media print {
    background: #fff;
    border: none;
    border-bottom: 2px solid #000;
    padding: 0.5rem 0;
    page-break-inside: avoid;

    > div {
      width: 33%;
      flex-grow: unset;
    }
  }
`;

export default function(props: CityProps & CityPropsFromDispatch) {
  const { weather, places, distance } = props;
  const placesProps = {
    props: { places: places.data },
    state: places.state,
  };
  const weatherProps = {
    props: { ...weather.data },
    state: weather.state,
  };
  return (
    <City>
      <CityMeta
        name={props.name}
        distance={distance.data}
        onMoveButtonClick={props.mover}
        onRemoveButtonClick={props.destroyer}
      />
      <Widget props={weatherProps} component={Weather} />
      <Widget props={placesProps} component={Places} />
    </City>
  );
}
