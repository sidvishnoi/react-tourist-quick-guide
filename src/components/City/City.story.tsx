import { action } from '@storybook/addon-actions';
import { number, select, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';

import * as React from 'react';
import City, { CityProps } from '.';

const stories = storiesOf('Organisms/City', module);
stories.addDecorator(withKnobs);

stories.add('city', () => {
  const props: CityProps = {
    distance: {
      data: number('distance', 50),
      state: 'ready',
    },
    name: text('city.name', 'Hyderabad'),
    places: {
      data: [
        {
          link: '#',
          summary: 'Iconic mosque with 4 minarets & arches',
          title: 'Charminar',
        },
        {
          link: '#',
          summary: 'Fortress complex famed for acoustics',
          title: 'Golconda Fort',
        },
        {
          link: '#',
          summary: 'Varied art & antiques from Asia & Europe',
          title: 'Salar Jung Museum',
        },
        {
          link: '#',
          summary: 'Heart-shaped lake with a Buddha statue',
          title: 'Hussain Sagar',
        },
      ],
      state: 'ready',
    },
    weather: {
      data: {
        forecast: [
          {
            icon: 113,
            summary: 'sunny',
            temperature: 30,
          },
          {
            icon: 119,
            summary: 'shower',
            temperature: 35,
          },
          {
            icon: 179,
            summary: 'sunny',
            temperature: 32,
          },
        ],
        icon: 260,
        summary: 'sunny',
        temperature: number('weather.temperature', 30, {
          max: 100,
          min: -100,
          range: true,
          step: 1,
        }),
      },
      state: 'ready',
    },
  };
  return (
    <City
      name={props.name}
      distance={props.distance}
      weather={props.weather}
      places={props.places}
      destroyer={action('REMOVE_CITY')}
      mover={action('MOVE_CITY_UP')}
    />
  );
});
