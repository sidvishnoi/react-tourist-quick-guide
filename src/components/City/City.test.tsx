import { mount } from 'enzyme';
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import React = require('react');
import * as renderer from 'react-test-renderer';
Enzyme.configure({ adapter: new Adapter() });

import City, { CityProps } from '.';

const cityProps: CityProps = {
  distance: {
    data: 23,
    state: 'ready',
  },
  name: 'CITY',
  places: {
    data: [],
    state: 'ready',
  },
  weather: {
    data: {
      forecast: [
        {
          icon: 'sun',
          summary: 'sunny',
          temperature: 39,
        },
        {
          icon: 'rain',
          summary: 'shower',
          temperature: 30,
        },
        {
          icon: 'sun',
          summary: 'sunny',
          temperature: 34,
        },
      ],
      icon: 'sun',
      summary: 'sunny',
      temperature: 40,
    },
    state: 'ready',
  },
};

describe('Component - City', () => {
  it('renders city with proper widgets', () => {
    const mover = (): null => null;
    const destroyer = (): null => null;

    const tree = renderer.create(
      <City mover={mover} destroyer={destroyer} {...cityProps} />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders city with widgets in loading state', () => {
    const mover = (): null => null;
    const destroyer = (): null => null;
    const props: CityProps = {
      distance: {
        data: null,
        state: 'loading',
      },
      name: 'CITY',
      places: {
        data: null,
        state: 'loading',
      },
      weather: {
        data: null,
        state: 'loading',
      },
    };

    const tree = renderer.create(
      <City mover={mover} destroyer={destroyer} {...props} />,
    );
    expect(tree).toMatchSnapshot();
  });
});
