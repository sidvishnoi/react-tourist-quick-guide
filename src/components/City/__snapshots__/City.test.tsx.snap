// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - City renders city with proper widgets 1`] = `
<div
  className="sc-iwsKbI dSEQiH"
>
  <div
    className="sc-gZMcBi dSwc"
  >
    <div
      className="buttons"
    >
      <button
        className="move sc-gqjmRU iKMjOf"
        onClick={[Function]}
        title="Move destination up in list"
      >
        ⬆
      </button>
      <button
        className="remove sc-gqjmRU iKMjOf"
        onClick={[Function]}
        title="Remove from list"
      >
        ✗
      </button>
    </div>
    <h3
      className="sc-VigVT hBMUxy"
    >
      CITY
    </h3>
    <div
      className="sc-jTzLTM gTioxh"
    >
      23
       KM
    </div>
  </div>
  <div
    className="sc-htoDjs lbxQv"
  >
    <div
      className="sc-bZQynM iuqrgX"
    >
      <div
        className="weather-today"
      >
        <div
          className="weather-day sc-EHOje iahMDE"
        >
          <span>
            40
          </span>
           °
          C
          <p>
            sunny
          </p>
        </div>
      </div>
      <div
        className="weather-forecast sc-gzVnrw hKGbEO"
      >
        <div
          className="weather-day sc-EHOje kdmaEr"
        >
          <span>
            39
          </span>
           °
          C
          <p>
            sunny
          </p>
        </div>
        <div
          className="weather-day sc-EHOje kdmaEr"
        >
          <span>
            30
          </span>
           °
          C
          <p>
            shower
          </p>
        </div>
        <div
          className="weather-day sc-EHOje kdmaEr"
        >
          <span>
            34
          </span>
           °
          C
          <p>
            sunny
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-htoDjs lbxQv"
  >
    <div
      className="sc-bdVaJa bDWFJH"
    >
      <h3
        className="sc-bwzfXH ibkAcI"
      >
        Top 0 tourist places:
      </h3>
      <ul
        className="sc-htpNat dKIRsi"
      />
    </div>
  </div>
</div>
`;

exports[`Component - City renders city with widgets in loading state 1`] = `
<div
  className="sc-iwsKbI dSEQiH"
>
  <div
    className="sc-gZMcBi dSwc"
  >
    <div
      className="buttons"
    >
      <button
        className="move sc-gqjmRU iKMjOf"
        onClick={[Function]}
        title="Move destination up in list"
      >
        ⬆
      </button>
      <button
        className="remove sc-gqjmRU iKMjOf"
        onClick={[Function]}
        title="Remove from list"
      >
        ✗
      </button>
    </div>
    <h3
      className="sc-VigVT hBMUxy"
    >
      CITY
    </h3>
    <div
      className="sc-jTzLTM gTioxh"
    >
      ...
    </div>
  </div>
  <div
    className="sc-htoDjs sc-dnqmqq cFffhV"
  >
    <svg
      className="loader"
      height="50"
      stroke="#ff8a00"
      viewBox="0 0 38 38"
      width="50"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fillRule="evenodd"
      >
        <g
          strokeWidth="2"
          transform="translate(1 1)"
        >
          <circle
            cx="18"
            cy="18"
            r="18"
            strokeOpacity=".5"
          />
          <path
            d="M36 18c0-9.94-8.06-18-18-18"
          >
            <animateTransform
              attributeName="transform"
              dur="1s"
              from="0 18 18"
              repeatCount="indefinite"
              to="360 18 18"
              type="rotate"
            />
          </path>
        </g>
      </g>
    </svg>
  </div>
  <div
    className="sc-htoDjs sc-dnqmqq cFffhV"
  >
    <svg
      className="loader"
      height="50"
      stroke="#ff8a00"
      viewBox="0 0 38 38"
      width="50"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fillRule="evenodd"
      >
        <g
          strokeWidth="2"
          transform="translate(1 1)"
        >
          <circle
            cx="18"
            cy="18"
            r="18"
            strokeOpacity=".5"
          />
          <path
            d="M36 18c0-9.94-8.06-18-18-18"
          >
            <animateTransform
              attributeName="transform"
              dur="1s"
              from="0 18 18"
              repeatCount="indefinite"
              to="360 18 18"
              type="rotate"
            />
          </path>
        </g>
      </g>
    </svg>
  </div>
</div>
`;
